File: projectEuler/7/7.cpp
//What is the 10001st prime number?
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

// Function to find the nth prime number
int nthPrime(int n) {
    // 2 is the only even prime number
    vector<int> primes;
    primes.push_back(2);
    int num = 3;
    while (primes.size() < n) {
        // Check if num is divisible by any prime number less than its square root
        bool isPrime = true;
        for (int prime : primes) {
            // If num is not divisible by any prime number less than its square root, it is a prime number
            if (prime > sqrt(num)) {
                break;
            }
            if (num % prime == 0) {
                // If num is divisible by any prime number less than its square root, it is not a prime number
                isPrime = false;
                break;
            }
        }
        if (isPrime) {
            // Add num to the list of prime numbers
            primes.push_back(num);
        }
        // Check only odd numbers
        num += 2;
    }
    // Return the nth prime number
    return primes.back();
}

int main() {
    cout << nthPrime(10001) << endl;
    return 0;
}

File: projectEuler/7/7.py
#What is the 10001st prime number?

# Returns True if n is prime, False otherwise
def isPrime(n):
    if n == 2:
        return True
    if n < 2 or n % 2 == 0:
        return False
    # Check divisibility by odd numbers up to the square root of n
    for i in range(3, int(n**0.5)+1, 2):
        # If n is divisible by i, it is not prime
        if n % i == 0:
            return False
    return True

def nthPrime(n):
    primes = 0
    i = 2
    while primes < n:
        if isPrime(i):
            # Increment the number of primes found
            primes += 1
        # If the number of primes found is equal to n, return i
        i += 1
    return i-1

print(nthPrime(10001))

